Enable SVT-AV1 mp4 and mkv/webm container support

TangZhiZhen opened the pull request:
Signed-off-by: ZhiZhen Tang <zhizhen.tang@intel.com>

EwoutH commented:
Thanks a lot for this PR! FFmpeg 4.2 will branch soon, it would be amazing to get SVT-AV1 integrated before the split.

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+    encode_sps_av1(bitstreamPtr, sequenceControlSetPtr);
+
+    // Copy SPS to the Output Bitstream
+    copy_rbsp_bitstream_to_payload(
```
Why we need this?
mp4 and webm are workable even we did not call this function.

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+
+    // Output buffer Allocation
+    EB_MALLOC(EbBufferHeaderType*, outputStreamBuffer, sizeof(EbBufferHeaderType), EB_N_PTR);
+    EB_MALLOC(uint8_t*, outputStreamBuffer->p_buffer, sizeof(uint8_t) * PACKETIZATION_PROCESS_BUFFER_SIZE, EB_N_PTR);
```
who will release this memory?

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+    EB_MALLOC(Bitstream*, bitstreamPtr, sizeof(Bitstream), EB_N_PTR);
+    EB_MALLOC(OutputBitstreamUnit*, bitstreamPtr->output_bitstream_ptr, sizeof(OutputBitstreamUnit), EB_N_PTR);
+
+    return_error = output_bitstream_unit_ctor(
```
Who will release the Bitstream? this function will allocate other 40M memory.

hassount replied:
this is allocated using an EB_MALLOC so it should be released at the de-init automatically.

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+        (uint32_t *)&(outputStreamBuffer->n_alloc_len),
+        encodeContextPtr);
+
+    *output_stream_ptr = outputStreamBuffer;
```
If we only allow user to call this function one time, we'd better write down it and add some check in this function.
If we allow user call this function multiple time, we'd better provide a way to free the memory. 
80+M each time is not a small number.

hassount replied:
every time this is called it's added to the memory table then freed at the de-init call

jsunintel commented on ffmpeg_plugin/0001-Add-ability-for-ffmpeg-to-run-svt-av1-with-svt-hevc.patch:
```diff
@@ -341,40 +343,52 @@ index 0000000000..a044c1eac5
-+ //   }
++    if (avctx->flags & AV_CODEC_FLAG_GLOBAL_HEADER) {
++        EbBufferHeaderType* headerPtr = (EbBufferHeaderType*)av_malloc(sizeof(EbBufferHeaderType));
++        if (!headerPtr) {
```
There is an EbBufferHeaderType object in svt_enc. Please use if instead of allocating one on the fly.

jsunintel commented on ffmpeg_plugin/0001-Add-ability-for-ffmpeg-to-run-svt-av1-with-svt-hevc.patch:
```diff
@@ -341,40 +343,52 @@ index 0000000000..a044c1eac5
++        if (!headerPtr->p_buffer) {
++            av_log(avctx, AV_LOG_ERROR,
++                   "Cannot allocate buffer size %ld.\n", sizeof(uint8_t) * PACKETIZATION_PROCESS_BUFFER_SIZE);
++            svt_ret = EB_ErrorInsufficientResources;
```
Such setting svt_ret is unnecessary, which is not used anywhere.

EwoutH commented:
@TangZhiZhen Could you resolve the conflicts? And are all discussions now resolved or implemented? Would be nice to merge! :)

xuguangxin commented:
HI @TangZhiZhen , We have a key issue in this patch.  eb_svt_enc_stream_header malloced many memories, but nobody can free it. 

output_bitstream_unit_ctor will allocate 40M  https://github.com/OpenVisualCloud/SVT-AV1/blob/2f4852a031277cfb7cfd2f07c7a4f41c8591fcc7/Source/Lib/Common/Codec/EbBitstreamUnit.c#L33
eb_svt_enc_stream_header will allocate other 2K+.
https://github.com/OpenVisualCloud/SVT-AV1/blob/28ac6217e5bcae0ff90e3270f6d5375dd5a2a4ca/Source/Lib/Encoder/Codec/EbEncHandle.c#L3202-L3203

Is it possible to get a empty buffer from output_stream_buffer_producer_fifo_ptr_dbl_array[0])[0]? 
Then user can call eb_svt_release_out_buffer  free it.

The prototype is here 
https://github.com/xuguangxin/SVT-AV1/commit/2d875194d128aad65518665548f260dbc84bd5d1
please take it as reference.

thanks

kelvinhu325 commented:
hi @xuguangxin ,  the output_stream_buffer_producer_fifo_ptr_dbl_array always appears with output_stream_buffer_consumer_fifo_ptr_dbl_array, looks not reasonable only use producer_fifo in ffmpeg, still suggest allocate outputStreamBuffer->p_buffer in ffmpeg, and EB_MALLOC in bitstreamPtr->output_bitstream_ptr for encode_sps_av1() is a common issue as others EB_MALLOC's issues.

hassount commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+    EbBufferHeaderType      *outputStreamBuffer;
+
+    // Output buffer Allocation
+    EB_MALLOC(EbBufferHeaderType*, outputStreamBuffer, sizeof(EbBufferHeaderType), EB_N_PTR);
```
just to be clear, with the current SVT structure no other SVT thread would be able to call EB_MALLOC at the same time. this part of SVT is single threaded and all buffers are allocated at init time. that said, the ffmpeg point is valid if the libraries are init in parallel

hassount commented on ffmpeg_plugin/0001-Add-ability-for-ffmpeg-to-run-svt-av1-with-svt-hevc.patch:
```diff
@@ -113,6 +113,8 @@ index 0000000000..a044c1eac5
 +#include "internal.h"
 +#include "avcodec.h"
 +
++#define PACKETIZATION_PROCESS_BUFFER_SIZE 40000000
```
this should not be known to ffmpeg not needed

TangZhiZhen replied:
yes, I have moved this marco to EbSvtAv1Enc.h

TangZhiZhen replied:
Done

xuguangxin commented:
Hi @hassount , 
thanks for comment.
>just to be clear, with the current SVT structure no other SVT thread would be able to call EB_MALLOC at the same time. this part of SVT is single threaded and all buffers are allocated at init time. that said, the ffmpeg point is valid if the libraries are init in parallel

multi encoders are very normal in ffmpeg. see https://trac.ffmpeg.org/wiki/Creating%20multiple%20outputs.  If we have multi encoders init in parallel., we will have a big problem.

>this is allocated using an EB_MALLOC so it should be released at the de-init automatically.

>every time this is called it's added to the memory table then freed at the de-init call

like I mentioned in https://github.com/OpenVisualCloud/SVT-AV1/issues/285#issuecomment-496787746  If we have 2 encoders, it's no way to decide who allocated the memory. So it's no way to free an encoder specific memory at an encoder deinit time.

From API level, if we give a buffer to caller, we'd better provide some release function to user.  Even for the current implementation,  the release function may be an empty function.
thanks

hassount commented:
@xuguangxin 
agreed, for both cases, is there a way to get this in first then change them when we address the general EB_MALLOC ? in which case the only AR is to add a release api like the eb_svt_release_out_buffer(&headerPtr);

xuguangxin commented:
@hassount , add new fuction may introduct API complexity.   Can we use  eb_svt_release_out_buffer to release eb_svt_enc_stream_header's output?  Actually, the stream header is an encoder output.  It's very reasonable to use eb_svt_release_out_buffer release it.
here is some prototype https://github.com/xuguangxin/SVT-AV1/commit/2d875194d128aad65518665548f260dbc84bd5d1

xuguangxin commented:
> hi @xuguangxin , the output_stream_buffer_producer_fifo_ptr_dbl_array always appears with output_stream_buffer_consumer_fifo_ptr_dbl_array, looks not reasonable only use producer_fifo in ffmpeg, still suggest allocate outputStreamBuffer->p_buffer in ffmpeg, and EB_MALLOC in bitstreamPtr->output_bitstream_ptr for encode_sps_av1() is a common issue as others EB_MALLOC's issues.

Hi @kelvinhu325 , 
Thanks for the reply. 
If we want to support multi-instances. We can't replay on deinit time free. Every component needs free the memory allocated by themselves. 
We generate a new buffer in  eb_svt_enc_stream_header, and give back to the user, we'd better provide a function to help user free the buffer.
thanks

TangZhiZhen commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3117,8 +3117,50 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+        PACKETIZATION_PROCESS_BUFFER_SIZE);
+
+    // Reset the bitstream before writing to it
+    reset_bitstream(
```
yes, I removed it

TangZhiZhen replied:
Done

TangZhiZhen commented on ffmpeg_plugin/0001-Add-ability-for-ffmpeg-to-run-svt-av1-with-svt-hevc.patch:
```diff
@@ -341,40 +343,52 @@ index 0000000000..a044c1eac5
-+ //       av_freep(&headerPtr->p_buffer);
-+ //   }
++    if (avctx->flags & AV_CODEC_FLAG_GLOBAL_HEADER) {
++        EbBufferHeaderType* headerPtr = (EbBufferHeaderType*)av_malloc(sizeof(EbBufferHeaderType));
```
yes , I will replace with a local variable

TangZhiZhen replied:
Done

TangZhiZhen commented:
> @TangZhiZhen Could you resolve the conflicts? And are all discussions now resolved or implemented? Would be nice to merge! :)

@hassount I have resolved the conflicts. I think all discussions are resolved.

xuguangxin commented:
> Hi @xuguangxin , EB_MALLOC calling in svt core will add a new buffer index, this new buffer could be released based on this buffer index when close svt core, as my understanding, multi encoder instances can generate different buffer indexes and will not cause releasing issue.

@kelvinhu325 , No, please see discussion here https://github.com/OpenVisualCloud/SVT-AV1/issues/285#issuecomment-496787746. 
Every eb_enc_handle_ctor will set its member variable enc_handle_ptr->memory_map to mem_map .  
mme_map is a global variable.  so if we two encoders encoder1,  then encoder2. if we destroy encoder1 before encoder2. 
All mem allocated by encoder2 will be released by encoder1. And all memory allocated by encoder1 will be leaked.

@TangZhiZhen , 
From API level, it's no memory leak now, You can merge it now.

But 
>#define PACKETIZATION_PROCESS_SPS_BUFFER_SIZE       2000.

is not an elegant solution,
1.  2000 is a magic number,  how the caller knows 2000 enough for AV1 sps?
2. SVT is codec family, we'd better define the same/simliar API for HEVC/VP9/AV1/....  How you know 2000 is enough for all codecs.
Actually, max sps size is an internal state. The caller does not need to know it.

thanks

hassount commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3163,8 +3163,46 @@ EB_API EbErrorType eb_svt_enc_stream_header(
+        (OutputBitstreamUnit*)bitstreamPtr->output_bitstream_ptr);
+
+    // Code the SPS
+    encode_sps_av1(bitstreamPtr, sequenceControlSetPtr);
```
Agreed, SVT-HEVC follows this pattern too, adds a configuration token to switch it ON vs OFF

xuguangxin replied:
@kelvinhu325, @hassount 
ffmpeg only call eb_svt_enc_stream_header  when the output is a [container format](https://github.com/OpenVisualCloud/SVT-AV1/pull/286/files#diff-280dcd3e39e9dfff92d2c1dc5e3a6387R344). 
If we skip this in the bitstream, ivf is not playable.

hassount replied:
then we add a flag to ask the library to encode it or not to like in svt-hevc: codeVpsSpsPps

hassount commented:
I'd hate to be nagging here but again this buffer size
```
#define PACKETIZATION_PROCESS_SPS_BUFFER_SIZE       2000
```
should not be used to allocate / defined anywhere close to the application or ffmpeg, this is an encoder issue. we would need to mimic what SVT-HEVC is doing here and add a flag to signal adding / removing the header from the stream if it's done at init time and last but not least, add an API to release the buffer.

xuguangxin commented:
@hassount , sorry for bringing this up again and again, it's my fault. 
But this is an important issue. ffmpeg community will ask the same question too :)

hassount commented:
> ffmpeg community will ask the same question too :)

@xuguangxin sorry I don't follow, can you elaborate a little more on what you are referring to ?

xuguangxin commented:
I mean PACKETIZATION_PROCESS_SPS_BUFFER_SIZE  should not know by ffmpeg. I first point it out at https://github.com/OpenVisualCloud/SVT-AV1/pull/286#discussion_r288417009 . and spend some comment to explain why it's bad. But after some review iterations, it back again. Although the name changed, and buffer size reduced from  40M to 2K. But the reality is not changed.

From your comment, I know you are aligned on this. So what is the next step? Merge this patch firstly or wait for new buffer header release function ready on this PR?

thanks

TangZhiZhen commented:
> I mean PACKETIZATION_PROCESS_SPS_BUFFER_SIZE should not know by ffmpeg. I first point it out at [#286 (comment)](https://github.com/OpenVisualCloud/SVT-AV1/pull/286#discussion_r288417009) . and spend some comment to explain why it's bad. But after some review iterations, it back again. Although the name changed, and buffer size reduced from 40M to 2K. But the reality is not changed.
> 
> From your comment, I know you are aligned on this. So what is the next step? Merge this patch firstly or wait for new buffer header release function ready on this PR?
> 
> thanks

Done, I have updated the code. Added an API to release the SPS buffer.

TangZhiZhen commented:
> I'd hate to be nagging here but again this buffer size
> 
> ```
> #define PACKETIZATION_PROCESS_SPS_BUFFER_SIZE       2000
> ```
> 
> should not be used to allocate / defined anywhere close to the application or ffmpeg, this is an encoder issue. we would need to mimic what SVT-HEVC is doing here and add a flag to signal adding / removing the header from the stream if it's done at init time and last but not least, add an API to release the buffer.

Done, I have updated the code. Added an API to release the SPS buffer. but I don't think we need add a flag to signal adding / removing the header from the stream. because  it has avctx->flags on ffmpeg, if avctx->flags = AV_CODEC_FLAG_GLOBAL_HEADER, we must call  eb_svt_enc_stream_header.

hassount commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3064,10 +3064,73 @@ __attribute__((visibility("default")))
-    EbErrorType             return_error = EB_ErrorNone;
-    UNUSED(svt_enc_component);
-    UNUSED(output_stream_ptr);
+    EbBufferHeaderType        **output_stream_ptr)
```
coding style EbBufferHeaderType        **output_stream_ptr){

hassount commented:
> but I don't think we need add a flag to signal adding / removing the header from the stream. because it has avctx->flags on ffmpeg, if avctx->flags = AV_CODEC_FLAG_GLOBAL_HEADER, we must call eb_svt_enc_stream_header.

So the way you have it implemented right now is the following: 
- get a header when muxing at init time
- then another header will be encoded by the encoder in the first picture in the EbPacketizationProcess.c file. 

So you'd end-up having two headers in the stream. so you'll need a way to ask the library to encode it or not to encode it depending on the need. 

if you decide to remove the encoded header from the library, you'll need to update the Svt application with such change.

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3064,10 +3064,73 @@ __attribute__((visibility("default")))
+    encode_sps_av1(&bitstream, sequenceControlSetPtr);
+
+    // Copy SPS to the Output Bitstream
+    copy_rbsp_bitstream_to_payload(
```
since you already allocate a buffer for outputStreamBuffer->p_buffer, can we just set ((OutputBitstreamUnit *)bitstream.output_bitstream_ptr)->buffer_begin_av1 to outputStreamBuffer->p_buffer?
it will
1. save one malloc and free.
2. we do not need call copy_rbsp_bitstream_to_payload to do useless copy

xuguangxin commented on Source/Lib/Encoder/Codec/EbEncHandle.c:
```diff
@@ -3064,10 +3064,73 @@ __attribute__((visibility("default")))
+    EbBufferHeaderType        *stream_header_ptr)
+{
+    EbErrorType           return_error = EB_ErrorNone;
+    free(stream_header_ptr->p_buffer);
```
better check stream_header_ptr for NULL

xuguangxin commented on Source/API/EbSvtAv1Enc.h:
```diff
@@ -457,6 +457,13 @@ typedef struct EbSvtAv1EncConfiguration
+     *
+     * Parameter:
+     * @ *stream_header_ptr  stream header buffer. */
+    EB_API EbErrorType eb_svt_enc_stream_header_dtor(
```
The function name is not so meanfull
Is it possbile follow same name tradition as eb_svt_release_out_buffer. 
for example eb_svt_release_enc_stream_header?

xuguangxin commented on Source/API/EbSvtAv1Enc.h:
```diff
@@ -457,6 +457,13 @@ typedef struct EbSvtAv1EncConfiguration
         EbComponentType           *svt_enc_component,
         EbBufferHeaderType       **output_stream_ptr);
 
+    /* OPTIONAL: Destroy stream headers at init time.
```
This function is not optional, a user who called eb_svt_enc_stream_header must call this to release memory.
also, a user can call this function anytime, not just init time

xuguangxin commented:
@TangZhiZhen, thanks for the update, seems we are close to finish it. 
But it still need some changes.
thanks

xuguangxin commented:
> So you'd end-up having two headers in the stream. so you'll need a way to ask the library to encode it or not to encode it depending on the need.

@hassount , No, it will not have two headers in bitstream. 
The sps generated by eb_svt_enc_stream_header will be copied to the avctx->extradata,  and send to muxer. The uuxer will parser sps and generate [AV1CodecConfigurationBox](https://aomediacodec.github.io/av1-isobmff/#av1codecconfigurationbox). The muxer will put the AV1CodecConfigurationBox to a container header, not the bitstream.

The IVF muxer does [not ](https://github.com/FFmpeg/FFmpeg/blob/master/libavformat/ivfenc.c)support extra data.  If we have ivf muxer, we need to send sps in bitstream.

Remove sps from bitstream will introduce API and caller complex. 
Any other strong reasons you want to remove sps from bitstream?
thanks

xuguangxin commented:
LGTM, thanks @TangZhiZhen addressed all issues.

tianjunwork commented:
Hi @TangZhiZhen, I tried mp4 and mkv, both work. WebM failed with below message. Could you take a look?
[webm @ 0x3067b40] Only VP8 or VP9 or AV1 video and Vorbis or Opus audio and WebVTT subtitles are supported for WebM.
Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument
Error initializing output stream 0:0 --
Conversion failed!

tianjunwork commented:
Hi @jsunintel, below command in ffmpeg plugin README doesn't work anymore. Could you update it to use the same command in SVT README?
cd Build && cmake .. && make -j nproc && sudo make install

TangZhiZhen commented:
> Hi @TangZhiZhen, I tried mp4 and mkv, both work. WebM failed with below message. Could you take a look?
> [webm @ 0x3067b40] Only VP8 or VP9 or AV1 video and Vorbis or Opus audio and WebVTT subtitles are supported for WebM.
> Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument
> Error initializing output stream 0:0 --
> Conversion failed!

Your input stream audio codec is Vorbis or Opus ?

TangZhiZhen commented:
> Hi @TangZhiZhen, I tried mp4 and mkv, both work. WebM failed with below message. Could you take a look?
> [webm @ 0x3067b40] Only VP8 or VP9 or AV1 video and Vorbis or Opus audio and WebVTT subtitles are supported for WebM.
> Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument
> Error initializing output stream 0:0 --
> Conversion failed!

Could you help to try this command?
ffmpeg -i input.mp4 -c:v libsvt_av1 -c:a libvorbis output-av1.webm

tianjunwork commented:
The audio track is ac3.
      handler_name    : GPAC ISO Video Handler
    Stream #0:1(und): Audio: ac3 (ac-3 / 0x332D6361), 48000 Hz, 5.1(side), fltp, 320 kb/s (default)

Tried with "-c:v libsvt_av1 -c:a **vorbis** -strict -2 output-av1.webm", Got vorbis error:
[vorbis @ 0x397b7c0] Current FFmpeg Vorbis encoder only supports 2 channels.
Error initializing output stream 0:1 -- Error while opening encoder for output stream #0:1 - maybe incorrect parameters such as bit_rate, rate, width or height.

If webm only accept vobis or Opus, please specify it in plugin README, instead of letting others fail.
Thank you.

BlauerHunger commented:
I can't get ffmpeg to build with this, neither the pr applied on top of the current master, nor using TangZhiZhen:upstream directly. The error I get is the following:

```
/usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
collect2: error: ld returned 1 exit status
make: *** [Makefile:108: ffplay_g] Error 1
make: *** Waiting for unfinished jobs....
/usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
collect2: error: ld returned 1 exit status
make: *** [Makefile:108: ffprobe_g] Error 1
/usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
collect2: error: ld returned 1 exit status
make: *** [Makefile:108: ffmpeg_g] Error 1
```

TangZhiZhen commented:
it looks like 

> I can't get ffmpeg to build with this, neither the pr applied on top of the current master, nor using TangZhiZhen:upstream directly. The error I get is the following:
> 
> ```
> /usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:108: ffplay_g] Error 1
> make: *** Waiting for unfinished jobs....
> /usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:108: ffprobe_g] Error 1
> /usr/bin/ld: libavcodec/libavcodec.so: undefined reference to `eb_svt_release_enc_stream_header'
> collect2: error: ld returned 1 exit status
> make: *** [Makefile:108: ffmpeg_g] Error 1
> ```

please updated the libSvtAv1Enc.so and EbSvtAv1Enc.h

BlauerHunger commented:
Now it works. Apparently I had to delete the installed files manually before running `make install` on SVT-AV1, even though it said installing … on the files in question. But now it builds

hassount commented:
ok are we all good with this? 
@kylophone  any issues with this ?

xuguangxin commented:
@BlauerHunger , it's related to https://github.com/OpenVisualCloud/SVT-AV1/blob/911599df35d01760fdae18c68b75aa9df3ac7cb9/ffmpeg_plugin/0001-Add-ability-for-ffmpeg-to-run-svt-av1.patch#L276, it will set intra_refresh_type to 1(CRA_REFRESH), but actually, av1 has no concept like CRA_REFRESH.  Except for first frame, every frame marked as a non-key frame.
Change the line to 2 (IDR_REFRESH) will fix your problem.

BlauerHunger commented:
@xuguangxin thanks, that did the job. 

What is the reason for it being set to `1` instead of `2 ` when this is a known problem?

xuguangxin commented:
@BlauerHunger , Some code copied from SVT HEVC, it's not clean yet. 
This issue has been blocked so long, how about we change param->intra_refresh_type  to 2, merge it find a graceful fix later?

@hassount , @kelvinhu325 , @TangZhiZhen , What do you think?

kelvinhu325 commented:
> @BlauerHunger , Some code copied from SVT HEVC, it's not clean yet.
> This issue has been blocked so long, how about we change param->intra_refresh_type to 2, merge it find a graceful fix later?
> 
> @hassount , @kelvinhu325 , @TangZhiZhen , What do you think?

Hi @xuguangxin, looks like CRA_REFRESH is another new issue and not closely relative with SVT-AV1 mp4 enabling change, to be more clear, how about creating an new PR for CRA_REFRESH fixing after this change is merged.

xuguangxin commented:
@kelvinhu325 , good to me too.

Please take  CRA_REFRESH thing as a high priority. Without it, the stream is not seekable.
thanks

kelvinhu325 commented:
> @kelvinhu325 , good to me too.
> 
> Please take CRA_REFRESH thing as a high priority. Without it, the stream is not seekable.
> thanks

@xuguangxin , ok, then zhizhen or who will be owner for CRA_REFRESH issue?

xuguangxin commented:
@kelvinhu325 , please assign to @zhizhen, 
I still busy with https://github.com/OpenVisualCloud/SVT-AV1/pull/314

TangZhiZhen commented:
> @kelvinhu325 , please assign to @zhizhen,
> I still busy with #314

@xuguangxin @kelvinhu325 ok, I will submit another pull request to resolve intra_refresh_type issue on SVT-AV1 FFmpeg plugin.

kelvinhu325 commented:
> > @kelvinhu325 , please assign to @zhizhen,
> > I still busy with #314
> 
> @xuguangxin @kelvinhu325 ok, I will submit another pull request to resolve intra_refresh_type issue on SVT-AV1 FFmpeg plugin.

ok, thanks @xuguangxin , Hi @hassount , if no any other issue, could you help merge this change now?

hassount commented:
@kylophone I am merging this assuming no objections.
